### ISSUE TITLE
in-place operator "+=" breaks autograd

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#10586]
## Issue description

We found an autograd issue with in-place operator "+=" when a custom layer is present.
Looks like the autograd gets cut off after the in-place operator.

## Code example

Runtime issue:
only forward function of AddKernel is triggered.
 
repro code:

```
  1 import torch
  2
  3 '''
  4 The following model breaks autograd.
  5
  6 During execution, we run both forward (LINE:71) and backward (LINE:73)
  7 we observe only the forward pass of our custom layer has been executed, while
  8 the backward pass does not gets triggered by.
  9 '''
 10
 11 class AddKernel(torch.autograd.Function):
 12     @staticmethod
 13     def forward(ctx, input, val):
 14         print("forward")
 15         out = input.transpose(-1, 1)
 16         ctx.save_for_backward(input, val)
 17         out = out.add(val)
 18         # For some reason, issue 1 only apears when we transpose the clone of a
 19         # tensor in this very specific way
 20         # Swapping it with any of the commented line below could make autograd
 21         # functioning properly again.
 22         out = out.clone().transpose(-1, 1)   #busted
 23         #out = out.clone().transpose_(-1, 1) #working fine
 24         #out = out.transpose(-1, 1).clone()  #working fine
 25         return out
 26
 27     @staticmethod
 28     def backward(ctx, grad_output):
 29         print("backward")
 30         inp, val = ctx.saved_tensors
 31         grad_input = torch.rand_like(inp)
 32         grad_val = torch.rand_like(val)
 33         return grad_input, grad_val
 34
 35
 36 class AddLayer(torch.nn.Module):
 37     def __init__(self, num_features):
 38         super(AddLayer, self).__init__()
 39         self.weight = torch.nn.parameter.Parameter(torch.Tensor(num_features))
 40     def forward(self, input):
 41         return AddKernel.apply(input, self.weight)
 42
 43
 44 class ResidualAddLayer(torch.nn.Module):
 45     def __init__(self):
 46         super(ResidualAddLayer, self).__init__()
 47         layers = []
 48         layers.append(torch.nn.Sequential(
 49             AddLayer(10),
 50         ))
 51         self.model = torch.nn.Sequential(*layers)
 52
 53     def forward(self, x):
 54         residual = x
 55         out = self.model(x)
 56         # issue 1: in-place operator '+=' used here cut off autograd tracing
 57         #          at run time, only AddKernel::forward function is executed.
 58         # swap it with the commented line below would work fine.
 59         # At run time, both AddKernel::forward & AddKernel::backward would be
 60         # executed.
 61         out += residual       # busted
 62         #out = out + residual # working fine
 63         return out
 64
 65 dtype = torch.float
 66 device = torch.device("cpu")
 67 N, C, H, W= 5, 10, 10, 10
 68 x = torch.randn(N, C, H, W, device=device, dtype=dtype)
 69 model = ResidualAddLayer()
 70 optimizer = torch.optim.SGD(model.parameters(), lr = 1e-1, momentum=0.9)
 71 y_pred = model(x)
 72 loss = y_pred - 1.0
 73 loss.backward(loss)
```

## System Info

Collecting environment information...
PyTorch version: 0.5.0a0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: Tesla V100-DGXS-16GB
GPU 1: Tesla V100-DGXS-16GB
GPU 2: Tesla V100-DGXS-16GB
GPU 3: Tesla V100-DGXS-16GB

Versions of relevant libraries:
[pip] msgpack-numpy (0.4.3)
[pip] numpy (1.15.0)
[pip] torch (0.5.0a0)
[pip] torchtext (0.3.0)
[pip] torchvision (0.2.1)

- PyTorch or Caffe2:
PyTorch

cc @ezyang @SsnL @albanD @zou3519

**Comment by jjsjann123** [comment]
I didn't fully understand the part `out` is treated as a view since the last operation is a viewing operation.
Does this mean that, as long as I have a viewing operation, any previous clone/detach would be ignored?

**Comment by albanD** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/issues/26546
